enhance: [2.4] Unify LoadStateLock RLock & PinIf (#39206)

congqixia opened the pull request:
Cherry-pick from master
pr: #39206 #39308
Related to #39205

This PR merge `RLock` & `PinIfNotReleased` into `PinIf` function preventing segment being released before any Read operation finished.

---------

congqixia committed:
enhance: Unify LoadStateLock RLock & PinIf (#39206)

Related to #39205

This PR merge `RLock` & `PinIfNotReleased` into `PinIf` function
preventing segment being released before any Read operation finished.

---------

Signed-off-by: Congqi Xia <congqi.xia@zilliz.com>

congqixia commented:
/run-cpu-e2e

congqixia committed:
Fix resolving conflict issue

Signed-off-by: Congqi Xia <congqi.xia@zilliz.com>

xiaofan-luan commented on internal/querynodev2/segments/state/load_state_lock.go:
```diff
@@ -129,76 +125,82 @@ func (ls *LoadStateLock) StartLoadData() (LoadStateLockGuard, error) {
+			// do nothing for empty segment.
+			g = nil
+		default:
+			panic(fmt.Sprintf("unreachable code: invalid state when releasing data, %s", ls.state.String()))
```
instead of panic, let's print a warn level log and return error

congqixia committed:
enhance: Log error instead of panicking if load lock wait timeout

Related to #39205
Previous PR #39206

This PR change wait timeout behavior to log error and return to avoid
making other collection read failure in only some collections have
deadlock

Signed-off-by: Congqi Xia <congqi.xia@zilliz.com>

congqixia committed:
Fix resolve unsafed

Signed-off-by: Congqi Xia <congqi.xia@zilliz.com>

xiaofan-luan commented:
/lgtm
/approve
